### ISSUE TITLE
NUT master-branch builds broken after recent PRs and/or infra changes

### ISSUE THREAD

**Issue by jimklimov** [networkupstools/nut#2249]
The two most-recent builds at https://ci.networkupstools.org/job/nut/job/nut/job/master/750/ and https://ci.networkupstools.org/job/nut/job/nut/job/master/751/ (after PRs #2245 and #2246) are broken in different ways.

* [ ] The first one (also discussed in PR comments) seems to have a problem with `AC_DISABLE_STATIC` in OpenIndiana builds. There are no relevant code changes, so possibly a regression on the NUT CI platform which during this month underwent (and will continue) hosting migration and upgrades. Possibly got some crippled autoconf version/half-installed package on one of the workers there?..

* The latter has numerous problems with other platforms as well, including older CentOS and cross-mingw builds, AppVeyor (mingw on windows), CircleCI (MacOS). Some are almost-benign complaints about coding style/static analysis in the newer compilers (e.g. requiring an explicit `funcname(void)` for lack of args), others are more serious with use of absent struct fields. @aquette : can you please take a look at those, if the code can be made portable or maybe `#ifdef` the platform-specific bits to they are not implemented in binary on incompatible platforms? Notably, this is what the CI-augmented PR mechanism is for, to not let these bugs creep into master ;) Alas, this time the real problem probably got obscured by the "expected-red" former issue.

  * [x] https://ci.networkupstools.org/job/nut/job/nut/job/master/751/artifact/.ci.MD5_d7ec2c323b3b230ceac0048bf1b93130.check.log.gz : Building with GCC STD=c99 STD=c++11 WARN=hard on i686-w64-mingw32 32-bit mingw-ubuntu-impish platform for MATRIX_TAG="c99-c++11-gcc-ubuntu-impish-i686-w64-mingw32-32bit-mayFail" && (ARCH_BITS=32&&ARCH32=i686-w64-mingw32&&COMPILER=GCC&&OS_DISTRO=ubuntu-impish&&OS_FAMILY=mingw) && (cross-windows-nut-builder) && (NUT_BUILD_CAPS=cross-windows-mingw) && BITS=32&&CSTDVARIANT=c&&CSTDVERSION_c=99&&CSTDVERSION_cxx=11 && LANG=C && LC_ALL=C && TZ=UTC && BUILD_TYPE=cross-windows-mingw && BUILD_WARNFATAL=yes && BUILD_WARNOPT=hard (isAllowedFailure) :: as part of slowBuild filter: Strict C and GNU standard builds on cross-Windows platforms (Linux+mingw), without distcheck and docs (allowed to fail): FAILED 'Test1'
````
../../../../../tools/nut-scanner/scan_nut_simulation.c: In function 'filter_ext':
../../../../../tools/nut-scanner/scan_nut_simulation.c:47:8: error: 'const struct dirent' has no member named 'd_type'
   47 |  if(dir->d_type == DT_REG) { /* only deal with regular file */
      |        ^~
../../../../../tools/nut-scanner/scan_nut_simulation.c:47:20: error: 'DT_REG' undeclared (first use in this function)
   47 |  if(dir->d_type == DT_REG) { /* only deal with regular file */
      |                    ^~~~~~
../../../../../tools/nut-scanner/scan_nut_simulation.c:47:20: note: each undeclared identifier is reported only once for each function it appears in
../../../../../tools/nut-scanner/scan_nut_simulation.c: In function 'nutscan_scan_nut_simulation':
../../../../../tools/nut-scanner/scan_nut_simulation.c:72:6: error: implicit declaration of function 'scandir' [-Werror=implicit-function-declaration]
   72 |  n = scandir(CONFPATH, &namelist, filter_ext, alphasort);
      |      ^~~~~~~
../../../../../tools/nut-scanner/scan_nut_simulation.c:72:47: error: 'alphasort' undeclared (first use in this function)
   72 |  n = scandir(CONFPATH, &namelist, filter_ext, alphasort);
      |                                               ^~~~~~~~~
````

  * Half-way-similar to the above on some other platforms, e.g. OpenIndiana ("the whole world is not an x86 Linux box", heh?) BTW this log also means that not all OI workers are broken with autoconf at the moment:
````
scan_nut_simulation.c: In function 'filter_ext':
scan_nut_simulation.c:47:15: error: 'const struct dirent' has no member named 'd_type'
   47 |         if(dir->d_type == DT_REG) { /* only deal with regular file */
      |               ^~
scan_nut_simulation.c:47:27: error: 'DT_REG' undeclared (first use in this function)
   47 |         if(dir->d_type == DT_REG) { /* only deal with regular file */
      |                           ^~~~~~
````

  * [x] https://app.circleci.com/pipelines/github/networkupstools/nut/552/workflows/eaef32b2-ecad-45b2-946d-10b9cb9a41e6/jobs/1586 :

````
./nut-scan.h:158:47: error: this function declaration is not a prototype [-Werror,-Wstrict-prototypes]
nutscan_device_t * nutscan_scan_nut_simulation();
                                              ^
                                               void
1 error generated.
````

**Comment by desertwitch** [comment]
I'm aware I'm not really entitled to an opinion here not having contributed all that much code-wise... but personally not really all that happy with the fast-paced pushing to the master at the moment, perhaps we can go back to tested and CI-approved pull requests as an ideal standard for keeping such bugs making their way into master on a minimum.

**Comment by desertwitch** [comment]
On another note I think we should stick to the principle of least surprise where possible as far as changes to crucial parts of the machinery are concerned. Perhaps also proposed changes to nut-scanner warrant some more discussion here (regarding dropping parameters, structural changes).

**Comment by aquette** [comment]
sorry for being pushy here. Wanted to see my new SmartNUT pet peeve coming to life fast, and the future on NUT main orientation (Linux more than OpenIndiana, though no underestimation) along... which I started long before retiring...
As a retired back to OSS a bit, I don't have anymore the whole chain (upstream-Debian...) in control, which is destabilizing :D
I agree with your PoV @desertwitch. that the end result (NUT release) must be clean! Sorry if it sounded disrespectful over @jimklimov's work on the CI. 

As an explanation: as the main useful NUT code producer (usbhid-ups, snmp-ups, dummy-ups, Debian package; nut-scanner, ...), despite retired, I'm still more, in favor of having 80% happy first and fast ;) A CI blocking for OpenIndiana over development is to me a concern! It should block for the release obviously, but not slow innovation. In Debian ,there are priorities on the ports.
That's how I made all this. I'm all for CI, it's the main part of my daily job. But let's not more royalist than the King ;)

**Comment by desertwitch** [comment]
My concern was towards the master branch, not necessarily only the releases. I think a lot of people have become used to building NUT from master and things being relatively stable and structured there (no real surprises due to Jim's very careful considerations and his principles of least surprise regarding changes). Also particularly thanks to what some may consider as a overly strict CI testing process before actual merging of new pull requests. This was further fueled by the relatively long gap between 2.8.0 and 2.8.1 releases, so a lot of people probably have become used to this project flow over the last two years - I myself have at least.

**Comment by aquette** [comment]
you're right about master. And the PR was scheduled. Just not the best timeframe...

So better act than say ;) That's again how I build (or help to) NUT and its many integrations and deployments

And just to be clear: these "premature PRs fast merged" are well working  / heavily tested code, improving NUT integration ability, and with extensive UT and real-life smoke tests. As I always did ;)
It just doesn't build on some venerable (not to say lagging behind the pack) systems. Which needs to be fixed. But with lower priority to me.

So again, better code than words to make NUT moving (and in the right direction), PRs is out and should fix... Otherwise, as usual, I'll take care.
Merry xmas to those celebrating today (back in a week @jimklimov  ;))

@desertwitch I encourage you to read about SmartNUT https://github.com/networkupstools/nut/projects?type=classic ;)
And possibly about the extent of my "previous contribs" ;) with all respect.
Sorry if it sounds impolite, I'm possibly too tired and not having fun (and recognition) coding opensource as before

**Comment by desertwitch** [comment]
@aquette Thanks for your response and perspective, it's much appreciated. I just wanted to put my personal opinion here (building and packaging on Slackware as one of the less commonly used OSes these days), with the utmost respect to your many contributions which shaped NUT into the project it is today - hell, my UPS wouldn't even have a driver without your efforts. And my opinion is just that, a single user's opinion, so do take it or leave it. Happy holidays and thanks again!! 😄

**Comment by aquette** [comment]
All apologies again if I sound disrespectful, too tired, too... too many things :D
Slackware was my 1rst Linux, back in 1995. It made me fall in Love with Free Software!
All due thanks and respect for your work :)
And be assured that I won't bullsh*t NUT after soooo many 1000s hours spent to make what it is :)

Again, take look at SmartNUT, despite just starting to (re)shape it, you should understand why this idea is exciting me as my 1rst Slackware, first contrib, first time I met NUT, ... ;)
@jimklimov can say he saw me doing that once, 15 years ago. But the landscape was not ready. Now it is...
And if you're interested in, I always need help to make my ideas come to life faster :D not just coding, any help

**Comment by jimklimov** [comment]
I think I found the OI issue - the migrated NUT CI controller pinged back to the original VM while it's alive to use as a build node, but got into a wrong container's ssh port.
